FWCore/ParameterSet/bin/edmWriteConfigs.cpp

```diff
@@ -136,18 +138,33 @@ namespace {
     }
   }
 
-  void writeModulesFile() {
+  bool open_temp(std::string& path, std::ofstream& f) {
+    path += "/XXXXXX";
+    path += '\0';  // ensure that the string is null-terminated
+    int fd = mkstemp(path.data());
+    if (fd != -1) {
+      f.open(path.c_str(), std::ios_base::trunc | std::ios_base::out);
+      close(fd);
+      return true;
+    }
+    return false;
+  }
```

Comment:
It would be better to separate the generation of a unique temporary file name and opening of the file. It would also be better to use the exceptions for error handling (since the caller of `writeModulesFile()` already has an exception handler).

Also, from C++11 onwards the `std::string::data()` is guaranteed to be null-terminated ([cppreference](https://en.cppreference.com/w/cpp/string/basic_string/data))

Please change along
```suggestion
  std::string makeUniqueFilename(std::string path) {
    path += "/XXXXXX";
    int fd = mkstemp(path.data());
    if (fd == -1) {
      throw cms::Exception("Failed unique file name") << "Unable to create a unique file name with the template " << path << ": " << std::strerror(errno);
    }
    close(fd);
    return path;
  }
```

Reply:
This approach (using a call to generate the file name, and a separate call to open the file) is not safe: that's the whole point of _not_ using `tmpnam()` in the first place.

Reply:
> This approach (using a call to generate the file name, and a separate call to open the file) is not safe

Why not? `mkstemp()` [documentation](https://pubs.opengroup.org/onlinepubs/9699919799/functions/mkdtemp.html) has
> The characters shall be chosen such that the resulting pathname does not duplicate the name of an existing file at the time of the call to `mkstemp()`

so the file name should be unique among other processes (which is better than the behavior of [`tmpnam()`](https://en.cppreference.com/w/cpp/io/c/tmpnam)
> it is possible that a file with that name is created by another process between the moment `std::tmpnam` returns and the moment this program attempts to use the returned name to create a file.

).

I agree it is possible that something else could tamper with the created file between the `close(fd)` and the `ofstream` construction. If this something else would be by us, we can change it (and currently there should be nothing else doing that). If this something else would be a malicious third party, I think we'd have bigger problems elsewhere.

Reply:
No, this is exactly the behaviour of `tmpnam`.

With [`std::tmpnam()`](https://en.cppreference.com/w/cpp/io/c/tmpnam):
  - create a unique filename that does not name a currently existing file, and stores it in the character string pointed to by `filename`
  - it is possible that a file with that name is created by another process between the moment `std::tmpnam` returns and the moment this program attempts to use the returned name to create a file.

With `mkstemp()` closing the file and recreating
  - the `mkstemp()` function generates a unique temporary filename from template, creates and opens the file, and returns an open file descriptor for the file.
  - if we close the file descriptor and open a new file, it's possible that another process creates a file with that same name in the meantime.

So replacing `tmpnam` with `mkstemp`, closing the file and re-opening, we do not gain absolutely anything.

Reply:
I believe there can be a difference between `tmpnam()` and `mkstemp()`. With two or more processes using `tmpnam()`, I'd expect it to be possible two (or more) of the processes to get the same value from their `tmpnam()` calls. When all the processes use `mkstemp()`, this would not happen as all the processes would be guaranteed to get different names (as long as the temporary files were kept around), or the `mkstemp()` call would fail.

Written that, given that the best solution would be to move the generation of `modules.py` to scram (https://github.com/cms-sw/cmssw/pull/44314#issuecomment-1979190732), which could hopefully be done at the timescale of a few weeks, it doesn't seem worth to spend too much time in improving the behavior, even with respect to `tmpnam()`.